### ISSUE TITLE
Course 2: Introduction to Deep Learning

### ISSUE THREAD

**Issue by Alex1986ham** [Alex1986ham/data_lakehouse_with_spark#1]
**Developing a handwritten Digits Classifier with Pytorch**

Project Summary
As a machine learning engineer, you’ve gained familiarity with deep learning, a powerful tool for a number of tasks – not the least of which is computer vision. As part of a new product, you’ve been asked to prototype a system for optical character recognition (OCR) on handwritten characters. Since the team is still collecting samples of data, you’ve been tasked with providing a proof of concept on the MNIST database of handwritten digits, a task with very similar input and output.

In this project, you will be given a Jupyter Notebook to do all of your coding and written explanations. You will preprocess a dataset for handwritten digit recognition, build a neural network, then train and tune that neural network using your data.

Before you get started, please review the Project Rubric page ahead, and ensure to see the Jupyter Notebook on the Environments page to get started.

Submission Instructions
Please save your trained model in the workspace or upload a single file for review.
Instructions Summary
Step 1
Load the dataset from torchvision.datasets
Use transforms or other PyTorch methods to convert the data to tensors, normalize, and flatten the data.
Create a DataLoader for your dataset
Step 2
Visualize the dataset using the provided function and either:
Your training data loader and inverting any normalization and flattening
A second DataLoader without any normalization or flattening
Explore the size and shape of the data to get a sense of what your inputs look like naturally and after transformation. Provide a brief justification of any necessary preprocessing steps or why no preprocessing is needed.
Step 3
Using PyTorch, build a neural network to predict the class of each given input image
Create an optimizer to update your network’s weights
Use the training DataLoader to train your neural network
Step 4
Evaluate your neural network’s accuracy on the test set.
Tune your model hyperparameters and network architecture to improve your test set accuracy, achieving at least 90% accuracy on the test set.
Step 5
Use torch.save to save your trained model.

Section 1: Data Loading and Exploration:

Transform data for use in neural networks. => Data is preprocessed and converted to a tensor, either using the .ToTensor() transform from torchvision.transforms or simply manually with torch.Tensor.

Use DataLoader to input training data to the neural network. => A DataLoader object for both train and test sets has been created using the train and test sets loaded from torchvision.

Explore datasets and describe their properties to set and optimize neural network parameters. => Notebook contains code which shows the size and shape of the training and test data.

The provided function or some other method (e.g. plt.imshow) is used to print one or more images from the dataset

Provide a brief justification of any necessary preprocessing steps or why no preprocessing is needed. => The submission contains a justification of necessary preprocessing steps (e.g. flattening, converting to tensor, normalization) in a comment or (preferably) a markdown block.

Section 2: Model Design and Training:

Use PyTorch to build a neural network for image classification.=> A Model or Sequential class is created with at least two hidden layers and implements a forward method that outputs a prediction probability for each of the 10 classes using softmax.

Select an appropriate loss function for training a classification network. => A loss function that works for classification tasks is specified.

Define an optimizer to minimize loss function and update model parameters for improved accuracy. => Any optimizer from torch.optim is used to minimize the loss function.

Section 3: Model Testing and Evaluation:

Use DataLoader and the holdout set to test the accuracy of the model. => The test DataLoader is used to get predictions from the neural network and compare the predictions to the true labels.

Optimize model hyperparameters to achieve a desired level of accuracy. => Hyperparameters are modified to attempt to improve accuracy and the model achieves at least 90% classification accuracy.

Save trained model parameters for later use. => The torch.save() function is used to save the weights of the trained model.

Save your Work
Please save your work in the workspace or upload a single Zip file for review. Before you submit, please ensure that you fulfill all of the Project [Rubric(opens in a new tab)](https://review.udacity.com/#!/rubrics/4816/view) criteria.
